src/scopeck.rs

```diff
@@ -151,156 +151,148 @@ struct ScopeState<'a> {
     frames_out: Vec<Frame>,
 }
 
-fn push_diagnostic(state: &mut ScopeState, ix: StatementIndex, diag: Diagnostic) {
-    state.diagnostics.entry(ix).or_insert(Vec::new()).push(diag);
-}
+impl<'a> ScopeState<'a> {
+    fn push_diagnostic(&mut self, ix: StatementIndex, diag: Diagnostic) {
```

Comment:
My current preference is that methods should only be used for things that represent something, especially when there's an abstraction barrier involved.  ScopeState doesn't represent anything; Frame does, so it could legitimately have methods.